File: TP2/TP2/src/main/java/SistemaImobiliaria/Funcoes.java

package SistemaImobiliaria;
import java.util.ArrayList;
import java.util.StringTokenizer;
import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ListIterator;
public class Funcoes {
    
    //Construtor da classe Funcoes
    public Funcoes(){
    }
    //funcao que le o arquivo
    public void lerArquivo(ArrayList <Imovel> v){
        Scanner arq = null;
        try{
            arq = new Scanner(new File("../database_imoveis.txt"));
        }
        catch(FileNotFoundException e){
            System.err.println("Error opening File");
            System.exit(1);
        }
        int i = 1;
        //loop que le os dados do arquivo 
        while(arq.hasNext()){
            StringTokenizer tok = new StringTokenizer(arq.nextLine(), ";");
            ArrayList <String> s = new ArrayList <>();
            while(tok.hasMoreTokens()){
                s.add(tok.nextToken());
                
            }
            String[] arm = new String[s.size()];
            s.toArray(arm);
            
            //switch case para verificar o tipo de propriedade
            switch (arm[0]) {
                //caso seja um apartamento
                case "apartamento" ->                     {
                        Apartamento objeto = new Apartamento();
                        objeto.setId(i);
                        objeto.setValor(Double.parseDouble(arm[1]));
                        objeto.setProprietario(arm[2]);
                        objeto.setRua(arm[3]);
                        objeto.setBairro(arm[4]);
                        objeto.setCidade(arm[5]);
                        objeto.setNumero(Integer.parseInt(arm[6]));
                        objeto.setQuartos(Integer.parseInt(arm[7]));
                        objeto.setBanheiros(Integer.parseInt(arm[8]));
                        objeto.setAndar(Integer.parseInt(arm[9]));
                        objeto.setTaxa_condominio(Double.parseDouble(arm[10]));
                        objeto.setElevador(intToBool(Integer.parseInt(arm[11])));
                        objeto.setSacada(intToBool(Integer.parseInt(arm[12])));
                        v.add(objeto);
                    }
                //caso seja uma casa
                case "casa" ->                     {
                        Casa objeto = new Casa();
                        objeto.setId(i);
                        objeto.setValor(Double.parseDouble(arm[1]));
                        objeto.setProprietario(arm[2]);
                        objeto.setRua(arm[3]);
                        objeto.setBairro(arm[4]);
                        objeto.setCidade(arm[5]);
                        objeto.setNumero(Integer.parseInt(arm[6]));
                        objeto.setQuartos(Integer.parseInt(arm[7]));
                        objeto.setBanheiros(Integer.parseInt(arm[8]));
                        objeto.setAndares(Integer.parseInt(arm[9]));
                        objeto.setSala_jantar(intToBool(Integer.parseInt(arm[10])));
                        v.add(objeto);
                    }
                //caso seja uma chacara
                default ->                     {
                            Chacara objeto = new Chacara();
                            objeto.setId(i);
                            objeto.setValor(Double.parseDouble(arm[1]));
                            objeto.setProprietario(arm[2]);
                            objeto.setRua(arm[3]);
                            objeto.setBairro(arm[4]);
                            objeto.setCidade(arm[5]);
                            objeto.setNumero(Integer.parseInt(arm[6]));
                            objeto.setQuartos(Integer.parseInt(arm[7]));
                            objeto.setBanheiros(Integer.parseInt(arm[8]));
                            objeto.setSalao_festa(intToBool(Integer.parseInt(arm[9])));
                            objeto.setSalao_jogos(intToBool(Integer.parseInt(arm[10])));
                            objeto.setCampo_futebol(intToBool(Integer.parseInt(arm[11])));
                            objeto.setChurrasqueira(intToBool(Integer.parseInt(arm[12])));
                            objeto.setPiscina(intToBool(Integer.parseInt(arm[13])));
                            v.add(objeto);
                    }
            }
            i++;
        }
        arq.close(); //fecha o arquivo
    }
    
    
    //funcao que converte um valor inteiro em booleano
    public boolean intToBool(int i){
        return (i > 0);
    }
    
    //Funcao que busca se um proprietario eh dono de um dos imoveis
    public boolean buscaProprietario(ArrayList <Imovel> v, String nome){
        
        //Loop que percorre o ArrayList
        for(Imovel E : v){
            //utiliza o getter para pegar um proprietario 
            String prop = E.getProprietario();
            if(prop.equals(nome)){
                return true;
            }
        }

        return false;   
    }
     
     //Funcao que retorna um ArrayList contendo os imoveis menores que um valor passado
    public ArrayList <Imovel> buscaPorValor(ArrayList <Imovel> v, double valorPesquisa){
	
	ArrayList <Imovel> r = new ArrayList <>();
        //Loop que percorre o ArrayList
	for(var E : v){
                //utiliza o getter para pegar o valor do imovel
		double valorImovel = E.getValor();
                
                //Caso o valor do imovel seja menor que o valor passado para pesquisa
		//coloca-se o imovel no ArrayList que sera retornado
		if(valorPesquisa >= valorImovel)
			r.add(E);
	}
        
	return r;
        
    }
    
    //Funcao que retorna um ArrayList contendo os imoveis com um numero minimo de quartos
    public ArrayList <Imovel> buscaPorQuartos(ArrayList <Imovel> v, int valorPesquisa){
        
        ArrayList <Imovel> r = new ArrayList <>();
        
        //loop que percorre o ArrayList
        for(var E : v){
            //utiliza o getter para pegar o numero de quartos dos imoveis 
            int valorImovel = E.getQuartos();
            
            //Caso o numero de quartos do imovel seja maior ou igual ao valor passado para pesquisa
            //coloca-se o imovel no ArrayList que sera retornado
            if(valorPesquisa <= valorImovel)
               r.add(E);
        }

	return r;
    } 
    
    //Funcao que retorna todos os imoveis de uma cidade especifica
    public ArrayList <Imovel> buscarCidade(ArrayList <Imovel> v, String nome){
       
        ArrayList <Imovel> r = new ArrayList <>();
        
        //loop que percorre o ArrayList
        for(Imovel E : v){
            //Pegar cidade do imovel por meio de um getter
            String city = E.getCidade();
            //Caso seja a cidade especificada, colocar no ArrayList a ser retornado
            if(nome.equals(city)){
                r.add(E);
            }
        }
        //ordena o ArrayList pelo valor do imovel
        r.sort((a,b) -> a.compareTo(b));
        return r;
    }
    //Funcao que retorna uma List de Iteradores do tipo Imovel de acordo com um Proprietario
    public ListIterator <Imovel> buscarImovelPorProprietario(ArrayList <Imovel> v, String nome){

	ArrayList <Imovel> r = new ArrayList <>();
	for(var E : v){
            //Pegar o proprietario por meio de um getter
            String prop = E.getProprietario();
            //Caso seja o proprietario especificado, colocar o ITERADOR no ArrayList
            if(prop.equals(nome)){
                r.add(E);
            }
	}
	//Retorno eh uma List de iteradores
        ListIterator <Imovel> u = r.listIterator();
	return u;
    }

    //Funcao que retorna um ArrayList de um tipo de imovel ordenado de acordo com o valor
    public ArrayList <Imovel> buscarTipoOrdenado(ArrayList <Imovel> v, String valorPesquisa){
        
	ArrayList <Imovel> r = new ArrayList <>();
	//Loop que percorre o ArrayList
	for(Imovel E : v){
            //Switch case para verificar o tipo de propriedade
            switch (valorPesquisa) {
                case "Apartamento":
                    if(E instanceof Apartamento)
                        r.add(E);
                    break;
                case "Casa":
                    if(E instanceof Casa)
                        r.add(E);
                    break;
                case "Chacara":
                    if(E instanceof Chacara)
                        r.add(E);
                    break;
                default:
                    System.out.println("Tipo de imovel invalido");
                    return r;
            }
	}
	//Funcao que ordena o ArrayList
	r.sort((a,b) -> a.compareTo(b));
	return r;
    }
}

File: TP01/funcoes.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <cstdio>
#include <cstdlib>
#include <string>
#include <sstream>
#include <algorithm>
#include "imovel.h"
#include "apartamento.h"
#include "casa.h"
#include "chacara.h"
using namespace std;

//Funcao que le o arquivo
void lerArquivo(vector <Imovel*> &v){
    ifstream dados;
	dados.open("database_imoveis.txt");
	if(!dados.good()){
		cout << " ⚠️  ERRO NA ABERTURA DO ARQUIVO ⚠️ " << endl;
		exit(EXIT_FAILURE);
	}
	v.clear();
	string aux;
	vector<string> arm;
	arm.clear();
	int i = 1;
	//Loop que le os dados do arquivo
	while(getline(dados, aux)){
		istringstream iss(aux);
		string token;
		//Loop que separa as entradas de cada variavel da classe
		while (getline(iss, token, ';')){
			arm.push_back(token);
		}
		//Caso seja um apartamento
		if(arm[0] == "apartamento"){
			Apartamento *objeto = new Apartamento;
			objeto->setId(i);
			objeto->setValor(stof(arm[1]));
			objeto->setProprietario(arm[2]);
			objeto->setRua(arm[3]);
			objeto->setBairro(arm[4]);
			objeto->setCidade(arm[5]);
			objeto->setNumero(stoi(arm[6]));
			objeto->setQuartos(stoi(arm[7]));
			objeto->setBanheiros(stoi(arm[8]));
			objeto->setAndar(stoi(arm[9]));
			objeto->setTaxaCondominio(stof(arm[10]));
			objeto->setElevador(stoi(arm[11]));
			objeto->setSacada(stoi(arm[12]));
			v.push_back(objeto);
		}
		//Caso seja uma Casa
		if(arm[0] == "casa"){
			Casa *objeto = new Casa;
			objeto->setId(i);
			objeto->setValor(stof(arm[1]));
			objeto->setProprietario(arm[2]);
			objeto->setRua(arm[3]);
			objeto->setBairro(arm[4]);
			objeto->setCidade(arm[5]);
			objeto->setNumero(stoi(arm[6]));
			objeto->setQuartos(stoi(arm[7]));
			objeto->setBanheiros(stoi(arm[8]));
			objeto->setAndares(stoi(arm[9]));
			objeto->setSalaJantar(stoi(arm[10]));
			v.push_back(objeto);
		}
		//Caso seja uma chacara
		if(arm[0] == "chacara"){
			Chacara *objeto = new Chacara;
			objeto->setId(i);
			objeto->setValor(stof(arm[1]));
			objeto->setProprietario(arm[2]);
			objeto->setRua(arm[3]);
			objeto->setBairro(arm[4]);
			objeto->setCidade(arm[5]);
			objeto->setNumero(stoi(arm[6]));
			objeto->setQuartos(stoi(arm[7]));
			objeto->setBanheiros(stoi(arm[8]));
			objeto->setSalaoFesta(stoi(arm[9]));
			objeto->setSalaoJogos(stoi(arm[10]));
			objeto->setCampoFutebol(stoi(arm[11]));
			objeto->setChurrasqueira(stoi(arm[12]));
			objeto->setPiscina(stoi(arm[13]));
			v.push_back(objeto);
		}
		arm.clear();
		i++;
	}
}
//Funcao que imprime o menu de opcoes
void printMenu(){
	cout << "\t\u250F\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2513\n";
	cout << "\t\u2503\033[1mMENU DE OPCOES\033[0m\u2503\n";
	cout << "\t\u2517\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u2501\u251B" << endl;
	cout << "🗒️  Tem-se as seguintes opcoes 🗒️ \n" << endl;
    cout << "1 - Mostrar lista de imoveis disponiveis (Digite '1 -s' para salvar em um arquivo externo)" << endl;
	cout << "2 - Verificar proprietario de imovel" << endl;
    cout << "3 - Ver imoveis abaixo de um valor" << endl;
    cout << "4 - Ver imoveis por numeros de quartos" << endl;
    cout << "5 - Ver imoveis por tipo e ordenados por um valor" << endl;
    cout << "6 - Ver imoveis localizados em uma cidade" << endl;
    cout << "7 - Ver imoveis por proprietario" << endl;
    cout << "0 - Finalizar programa\n" << endl;
    cout << "▶️  Digite a opcao desejada: ";
}
//Funcao que imprime o conteudo do container vector
void ImprimeContainer(vector <Imovel*> v, bool arquivo){
	//Variavel booleana que fala se deve-se imprimir em um arquivo externo
    if(!arquivo){
        for(vector<Imovel*>:: iterator itr = v.begin(); itr != v.end(); ++itr){
			Chacara *ch;
			//downcasting
			ch = dynamic_cast <Chacara*>(*itr);
			if(ch != nullptr){
				cout << *ch << endl;
			}
			else{
				//downcasting
				Casa *ca;
				ca = dynamic_cast <Casa*>(*itr);
				if(ca != nullptr){
					cout << *ca << endl;
				}
				else{
					//downcasting
					Apartamento *auxiliar = dynamic_cast <Apartamento*>(*itr);
					cout << *auxiliar  << endl;
				}
			}		
		}
    }
    else{
		//abertura do arquivo externo
		ofstream saida;
		saida.open("saida.txt");
		for(vector<Imovel*>:: iterator itr = v.begin(); itr != v.end(); ++itr){
			//downcasting
			Chacara *ch;
			ch = dynamic_cast <Chacara*>(*itr);
			if(ch != nullptr){
				saida << *ch << endl;
			}
			else{
				//downcasting
				Casa *ca;
				ca = dynamic_cast <Casa*>(*itr);
				if(ca != nullptr){
					saida << *ca << endl;
				}
				else{
					//downcasting
					Apartamento *auxiliar = dynamic_cast <Apartamento*>(*itr);
					saida << *auxiliar  << endl;
				}
			}		
		}
	}
}
//Funcao que busca se um proprietario eh dono de um dos imoveis
bool buscaProprietario(vector <Imovel*> v){
	cout << "Digite o nome do proprietario a ser buscado: ";
	string nome;
	getline(cin, nome);
	//Loop que percorre o container
	for(vector<Imovel*>:: iterator itr = v.begin(); itr != v.end(); ++itr){
		//Pegar o proprietario por meio de um getter
		string prop = (*itr)->getProprietario();
		if(nome == prop)
			return true;
	}
	return false;
}
//Funcao que retorna um vector contendo os imoveis menor que um valor passado
vector <Imovel*> buscaPorValor(vector <Imovel*> v){
	cout << "Digite o valor maximo a ser pesquisado: ";
	float valorPesquisa;
	cin >> valorPesquisa;
	getchar();
	cout << endl;
	vector <Imovel*> r;
	//Loop que percorre o vector
	for(vector<Imovel*>:: iterator itr = v.begin(); itr != v.end(); ++itr){
		//Pegar o valor do imovel por meio de um getter
		float valorImovel = (*itr)->getValor();
		//Caso o valor do imovel seja menor que o valor passado para pesquisa
		//coloca-se o imovel no vector que sera retornado
		if(valorPesquisa >= valorImovel)
			r.push_back(*itr);
	}
	
	return r;
}
//Funcao que retorna um vector contendo os imoveis com um numero minimo de quartos
vector <Imovel*> buscaPorQuartos(vector <Imovel*> v){
	cout << "Digite o numero de quartos minimo a ser pesquisado: ";
	int valorPesquisa;
	cin >> valorPesquisa;
	getchar();
	cout << endl;
	vector <Imovel*> r;
	//Loop que percorre o vector
	for(vector<Imovel*>:: iterator itr = v.begin(); itr != v.end(); ++itr){
		//Pegar o numero de quartos dos imoveis por meio de um getter
		int valorImovel = (*itr)->getQuartos();
		//Caso o numero de quartos do imovel seja maior ou igual ao valor passado para pesquisa
		//coloca-se o imovel no vector que sera retornado
		if(valorPesquisa <= valorImovel)
			r.push_back(*itr);
	}

	return r;
}
//Funcao que retorna um vector de um tipo de imovel ordenado de acordo com o valor
vector <Imovel*> buscarTipoOrdenado(vector <Imovel*> v){
	cout << "Digite o tipo de imovel a ser pesquisado: ";
	string valorPesquisa;
	getline(cin, valorPesquisa);
	//Tratar a string
	for_each(valorPesquisa.begin(), valorPesquisa.end(), [](char &c){c = ::toupper(c);});
	vector <Imovel*> r;
	//Loop que percorre o vector
	for(vector<Imovel*>:: iterator itr = v.begin(); itr != v.end(); ++itr){
		//Caso a pesquisa seja por apartamentos
		if(valorPesquisa == "APARTAMENTO"){
			Apartamento *auxiliar = dynamic_cast <Apartamento*>(*itr);
			if(auxiliar != nullptr)
				r.push_back(*itr);
		}
		//Caso a pesquisa seja por casas
		else if(valorPesquisa == "CASA"){
			Casa *auxiliar = dynamic_cast <Casa*>(*itr);
			if(auxiliar != nullptr)
				r.push_back(*itr);
		}
		//Caso a pesquisa seja por chacaras
		else if(valorPesquisa == "CHACARA"){
			Chacara *auxiliar = dynamic_cast <Chacara*>(*itr);
			if(auxiliar != nullptr)
				r.push_back(*itr);
		}
		//Caso tenha colocado um tipo invalido de imovel
		else{
			cout << "Tipo de imovel invalido" << endl;
			return r;
		}
	}
	//Funcao que ordena o vector
	sort(r.begin(), r.end(), [](Imovel* a, Imovel* b){return a->getValor() < b->getValor();});
	return r;
}
//Funcao que retorna todos os imoveis de uma cidade especifica
vector <Imovel*> buscarCidade(vector <Imovel*> v){
	cout << "Digite a cidade que deseja procurar por imoveis: ";
	string cidade;
	getline(cin, cidade);
	vector <Imovel*> r;
	//Loop que percorre o vector
	for(vector<Imovel*>:: iterator itr = v.begin(); itr != v.end(); ++itr){
		//Pegar cidade do imovel por meio de um getter
		string city = (*itr)->getCidade();
		//Caso seja a cidade especificada, colocar no vetor a ser retornado
		if(city == cidade){
			r.push_back(*itr);
		}
	}
	//Ordena o vector pelo valor do imovel
	sort(r.begin(), r.end(), [](Imovel* a, Imovel* b){ return a->getValor() > b->getValor();});
	return r;
}
//Funcao que retorna os imoveis de um proprietario
vector <vector <Imovel*> :: iterator > buscarImovelPorProprietario(vector <Imovel*> v){
	cout << "Digite o nome do proprietario a ser buscado: ";
	string nome;
	getline(cin, nome);
	vector <vector <Imovel*> :: iterator > r;
	//Percorre o vetor
	for(vector<Imovel*>:: iterator itr = v.begin(); itr != v.end(); ++itr){
		//Pegar o proprietario por meio de um getter
		string prop = (*itr)->getProprietario();
		//Caso seja o proprietario especificado, colocar o ITERADOR no vector
		if(prop == nome){
			r.push_back(itr);
		}
	}
	//Retorno eh um vetor de iteradores
	return r;
}
